### PR TITLE
feat: sui verify message handler

### PR DESCRIPTION
**Pull request by haiyizxx**
## Todos

- [x] Unit tests
- [x] Manual tests
- [ ] Documentation
- [x] Connect epics/issues

### CODE REVIEW

File: ampd/src/queue/queued_broadcaster.rs [hunk 1]
```diff
@@ -38,11 +39,12 @@ impl QueuedBroadcasterDriver {
     }
 }
 
+#[automock]
 #[async_trait]
 pub trait BroadcasterClient {
     async fn broadcast<T>(&self, tx: T) -> Result
     where
-        T: Msg + Send + Sync;
+        T: Msg + Send + Sync + 'static;
```
Comment (fish-sammy) [inline ampd/src/queue/queued_broadcaster.rs:47]: Having `'static` is a requirment from automock I assume?

File: ampd/src/sui/json_rpc.rs [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 use ethers::providers::{JsonRpcClient, ProviderError};
+use mockall::automock;
 use sui_json_rpc_types::{SuiTransactionBlockResponse, SuiTransactionBlockResponseOptions};
 use sui_types::digests::TransactionDigest;
 
 use crate::json_rpc::Client;
 
 type Result<T> = error_stack::Result<T, ProviderError>;
 
+#[automock]
```
Comment (fish-sammy) [inline ampd/src/sui/json_rpc.rs:11]: Get the method to return `Result<Vec<Option<SuiTransactionBlockResponse>>>` as discussed.

**Review by fish-sammy — CHANGES_REQUESTED** [review]

File: ampd/src/queue/queued_broadcaster.rs (hunk 1 above)
Comment (haiyizxx) [reply ampd/src/queue/queued_broadcaster.rs:47]: yep, I think it's fine to add the restriction for async

File: ampd/src/handlers/voter.rs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        assert!(!voter.is_voting_verifier(&TMAddress::random(PREFIX)));
+    }
+
+    #[async_test]
+    async fn should_submit_vote() {
+        let mut broadcaster = MockBroadcasterClient::new();
+
+        broadcaster
+            .expect_broadcast()
+            .once()
```
Comment (fish-sammy) [inline ampd/src/handlers/voter.rs:139]:
```suggestion
            .once().with(...)
```
to ensure it's voting for the correct thing.

File: ampd/src/sui/json_rpc.rs [hunk 4]
```diff
@@ -22,18 +26,24 @@ where
...
+                    .map(TransactionDigest::base58_encode)
                     .collect::<Vec<_>>(),
                 SuiTransactionBlockResponseOptions::new().with_events(),
             ),
         )
         .await
+        .map(|vec: Vec<SuiTransactionBlockResponse>| {
+            vec.into_iter()
+                // None checkpoint means the transaction is not finalized
+                .map(|block| block.checkpoint.map(|_| block))
```
Comment (fish-sammy) [inline ampd/src/sui/json_rpc.rs:45]: Their docs mention `"This is only returned in the read api, not in the transaction execution api"`. What's the difference between the read API and the execution API? Also, it seems that this `checkpoint` is only Some(...) when the transaction is included as well as finalized. Can you rename the method `transaction_blocks` to `finalized_transaction_blocks` then?

File: ampd/src/types.rs [hunk 5]
```diff
@@ -55,3 +55,21 @@ impl fmt::Display for TMAddress {
...
+    impl TMAddress {
+        pub fn random(prefix: &str) -> Self {
+            Self(
+                PublicKey::from(SigningKey::random(&mut OsRng).verifying_key())
+                    .account_id(prefix)
+                    .expect("failed to convert to account identifier"),
+            )
+        }
+    }
+}
```
Comment (fish-sammy) [inline ampd/src/types.rs:75]: @cgorenflo weigh in your opinion here. The point was that I don't like including code in the production build for it that is only needed by testing. Not sure if this is the best way/location to do this though.

File: ampd/src/sui/json_rpc.rs (hunk 4 above)
Comment (haiyizxx) [reply ampd/src/sui/json_rpc.rs]: sure good name. the execution api execute transaction to blockchain and wait for response

File: ampd/src/handlers/sui_verify_msg.rs [hunk 6]
```diff
@@ -12,3 +29,322 @@ pub struct Message {
...
+        rpc_client: C,
+        broadcast_client: B,
+    ) -> Self {
+        Self {
+            worker,
+            voting_verifier,
+            rpc_client,
+            broadcast_client,
+        }
+    }
```
Comment (cgorenflo) [inline ampd/src/handlers/sui_verify_msg.rs:73]: not necessary, it doesn't do any translation or checks, so you might as well use the object initializer

File: ampd/src/handlers/sui_verify_msg.rs [hunk 7]
```diff
@@ -12,3 +29,322 @@ pub struct Message {
...
+        }
+
+        let deduplicated_tx_ids: HashSet<_> = messages.iter().map(|msg| msg.tx_id).collect();
+        let transaction_blocks = self
+            .rpc_client
+            .finalized_transaction_blocks(deduplicated_tx_ids)
+            .await
+            .change_context(Error::TxReceipts)?;
+
+        let votes = messages
```
Comment (cgorenflo) [inline ampd/src/handlers/sui_verify_msg.rs:130]: add a comment that explains that the handler doesn't make any assumptions about id duplication in the event, but the rpc call fails(?) if there are duplicates, therefore you have to do this convolution of deduplication and mapping

**Review by cgorenflo — APPROVED** [review]

**Review by fish-sammy — APPROVED** [review]

**Pull request merged**